### ISSUE TITLE
join is very slow burning system time on seek()

### ISSUE THREAD

**Issue by jkleint** [BurntSushi/xsv#12]
I was comparing `xsv join` to GNU `join` and noticed that xsv seems to be doing random seeks on its input.  This is very slow for even moderate-sized files.  From the docs ("Uses a simple hash index to make it fast") I assumed it reads one file into memory and then streams through the other, but perhaps the design was meant to operate out-of-core?  (Explicitly building indexes does not help either.)

```
#!/bin/bash

export LC_ALL=C LINES=1000000 SORT_MEM=1G
seq $LINES | awk -vOFS=, "{print \$1, int(rand() * $LINES)}" > a.csv
seq $LINES | awk -vOFS=, "{print \$1, int(rand() * $LINES)}" > b.csv
time join -t, -j2 -o 1.1,1.2,2.1,2.2 <(sort -S"$SORT_MEM" -t, -k2,2 a.csv) <(sort -S"$SORT_MEM" -t, -k2,2 b.csv) | wc -l
#time (xsv index a.csv ; xsv index b.csv)
time xsv join -n 2 a.csv 2 b.csv | wc -l
rm -f a.csv b.csv a.csv.idx b.csv.idx
```

```
$ ./join.sh 
1999368

real	0m0.830s
user	0m0.480s
sys	0m0.076s
1999368

real	0m27.801s
user	0m3.004s
sys	0m24.836s
```
If it is meant to operate out-of-core, an option to operate in-core would be awesome.

**Comment by BurntSushi** [comment]
Yeah, the design was meant to be able to handle very large CSV data. It requires memory proportional to the cardinality of the set of join keys rather than just loading everything into CSV data.

As you've observed, this does not work so well when the CSV data can fit into memory.

The [random sampling](https://github.com/BurntSushi/xsv/blob/master/src/cmd/sample.rs) command would suffer from the same fate, but it's clever enough to only use random access when the sample is a small proportion of the data set.

Basically, we need a way to either take advantage of sequential IO or be more clever about loading CSV data into memory. The latter is probably easier. Might take me a bit to get to though.
